jstz_api/src/encoding/text_decoder.rs

```diff
@@ -0,0 +1,381 @@
+        accessor!(
+            context,
+            TextDecoder,
+            "encoding",
+            get:((x, context) => Ok(x.encoding.name().to_string().to_lowercase().into_js(context)))
+        )
+    }
+    fn fatal(context: &mut Context<'_>) -> Accessor {
+        accessor!(
+            context,
+            TextDecoder,
+            "fatal",
+            get:((x, _context) => Ok((x.error_mode == "fatal").into()))
+        )
+    }
+    fn ignore_bom(context: &mut Context<'_>) -> Accessor {
+        accessor!(
+            context,
+            TextDecoder,
+            "ignoreBOM",
+            get:((x, _context) => Ok(x.ignore_bom.into()))
+        )
+    }
+
+    //  https://encoding.spec.whatwg.org/#concept-td-serialize
+    fn serialize(&mut self, read: usize) -> () {
+        //  1. Let output be the empty string.
+        //  2. While true:
+        //    1. Let item be the result of reading from ioQueue.
+        //    2. If item is end-of-queue, then return output.
+        //    3. If decoder’s encoding is UTF-8 or UTF-16BE/LE, and decoder’s ignore BOM and BOM seen are false, then:
+        //      1. Set decoder’s BOM seen to true.
+        //      2. If item is U+FEFF, then continue.
+        //  4. Append item to output.
+
+        //  NB: we call this function in case of a chunked input with `read`:
+        //      https://docs.rs/encoding_rs/latest/encoding_rs/struct.Decoder.html
+        //  NB: BOM handling is done at the level of the decoding function from `::encoding_rs` crate
+        self.io_queue.drain(0..read);
+    }
+
+    //  https://encoding.spec.whatwg.org/#dom-textdecoder-decode
+    fn decode(
+        // input: Option<allowsharedbufersource>,
+        &mut self,
+        input: Option<Vec<u8>>,
+        options: Option<TextDecodeOptions>,
+    ) -> JsResult<String> {
+        //  Handle optional parameters
+        let input = input.unwrap_or(vec![]);
+        let options = options.unwrap_or(TextDecodeOptions::default());
+        //  1. If this's do not flush is false,
+        if !self.do_not_flush {
+            //  then set this's decoder to a new instance of this's encoding's decoder,
+            self.decoder = if self.ignore_bom {
+                self.encoding.new_decoder_without_bom_handling()
+            } else {
+                self.encoding.new_decoder()
+            };
+            //  this's I/O queue to the I/O queue of bytes <<end-of-queue>>,
+            self.io_queue.clear();
+            //  and this's BOM seen to false.
+            //  NB: we do not need this flag (intern to `encoding_rs` crate decode functions)
+        };
+        //  2. Set this's do not flush to options["stream"].
+        self.do_not_flush = options.stream;
+        //  3. If input is given, then push a copy of input to this's I/O queue.
+        //  FIXME: map this error into a JsError ? which one ?
+        let _: Result<usize, std::io::Error> = self.io_queue.write(input.as_slice());
```

Comment:
Can be something simple and uniquely identifiable for now
```suggestion
        self.io_queue.write(input.as_slice()).map_err(|_| JsNativeError::eval().with_message("IO error when writing to IO queue in TextDecoder").into())
```